setup-env

```diff
@@ -4,53 +4,201 @@ if [ -z "$ZSH_NAME"] && [ "x$0" = "x./setup-env" ]; then
+The second usage is for using an existing build directory. In this case,
+the script prepares the calling shell for running bitbake on the build
+directory <build-dir>. The build directory configuration is unchanged.
+"
+
+    ls layers/*/conf/machine/*.conf > /dev/null 2>&1
+    ls layers/meta-testdistro/conf/distro/*.conf > /dev/null 2>&1
+    if [ $? -eq 0 ]; then
+        echo -e "
+Supported machines: `echo; ls layers/*/conf/machine/*.conf \
+| sed s/\.conf//g | sed -r 's/^.+\///' | xargs -I% echo -e "\t%"`
+Supported Tegra test distros: `echo; ls layers/meta-testdistro/conf/distro/*.conf \
+| sed s/\.conf//g | sed -r 's/^.+\///' | xargs -I% echo -e "\t%"`
+Examples:
+- To create a new Yocto build directory:
+  $ MACHINE=jetson-tx2 DISTRO=testdistro source $PROGNAME build
+- To use an existing Yocto build directory:
+  $ source $PROGNAME build
+"
     fi
+}
+
+clean_up()
+{
+   unset LIST_MACHINES VALID_MACHINE
+   unset CWD TEMPLATES SHORTOPTS LONGOPTS ARGS PROGNAME
+   unset generated_config updated
+   unset MACHINE SDKMACHINE DISTRO OEROOT TEMPLATECONF BBPATH
+}
+
+# get command line options
+SHORTOPTS="h"
+LONGOPTS="help"
+
+ARGS=$(getopt --options $SHORTOPTS  \
+  --longoptions $LONGOPTS --name $PROGNAME -- "$@" )
+# Print the usage menu if invalid options are specified
+if [ $? != 0 -o $# -lt 1 ]; then
+   usage && clean_up
+   return 1
+fi
+
+eval set -- "$ARGS"
+while true;
+do
+    case $1 in
+        -h|--help)
+           usage
+           clean_up
+           return 0
+           ;;
+        --)
+           shift
+           break
+           ;;
+    esac
+done
+
+if [ "$(whoami)" = "root" ]; then
+    echo "ERROR: do not use the BSP as root. Exiting..."
+fi
+
+if [ ! -e $1/conf/local.conf ]; then
+    build_dir_setup_enabled="true"
 else
-    mkdir -p "$BDIR"
-    BDIR=`readlink -f "$BDIR"`
+    build_dir_setup_enabled="false"
+fi
+
+if [ "$build_dir_setup_enabled" = "true" ] && [ -z "$MACHINE" ]; then
+    usage
+    echo -e "ERROR: You must set MACHINE when creating a new build directory."
+    clean_up
+    return 1
 fi
+
+if [ -z "$SDKMACHINE" ]; then
+    SDKMACHINE='x86_64'
+fi
+
+if [ "$build_dir_setup_enabled" = "true" ] && [ -z "$DISTRO" ]; then
+    usage
+    echo -e "ERROR: You must set DISTRO when creating a new build directory."
+    clean_up
+    return 1
+fi
+
 # Allow for multiple layer sets,
 # default just "layers"
 if [ "x$BMETA" = "x" ]; then
     BMETA="layers"
 fi
-OEROOT=`readlink -f "$OEROOT/$BMETA"`
+OEROOT=`readlink -f "$CWD/$BMETA"`
 export OEROOT
-export BITBAKEDIR="$OEROOT/bitbake"
-. $OEROOT/scripts/oe-buildenv-internal && \
-    $OEROOT/scripts/oe-setup-builddir && \
-    [ -n "$BUILDDIR" ] && cd $BUILDDIR
-TEMPLATECONF=`cat $BUILDDIR/conf/templateconf.cfg`
+
+export TEMPLATECONF=${TEMPLATECONF:-$OEROOT/meta-testdistro/conf}
+
+. $OEROOT/openembedded-core/oe-init-build-env $CWD/$1 > /dev/null
+
+# if conf/local.conf not generated, no need to go further
+if [ ! -e conf/local.conf ]; then
```

Comment:
Shouldn't this be checking for `-n "$BUILDDIR" -a -e "$BUILDDIR/conf/local.conf"`?

Oh, I see you've changed to invoking the full `oe-init-build-env`, so I guess the current working directory will have been changed to $BUILDDIR if it's successful.  But if there's an error, and you happen to have a `conf/local.conf` under your original working directory... ?